code/modules/antagonists/traitor/smart_objectives.dm

```diff
@@ -0,0 +1,323 @@
+
+
+/**
+ * ## smart objectives
+ *
+ * Smart objectives are objectives for traitors that signal into their targets to confirm when the target dies
+ * This means they know when they are complete before roundend, and let an attached antag datum know
+ * (antag datums that care for smart objectives need to listen into them!)
+ */
+/datum/objective/smart
+	///this will be false the first time the objective is complete. if it is ever uncompleted, future completions will have this be TRUE
+	var/uncompleted = FALSE
+	///the payout for completing this objective
+	var/black_telecrystal_reward = 0
+
+/datum/objective/smart/proc/complete_objective()
+	if(completed)
+		return
+	completed = TRUE
+	SEND_SIGNAL(src, COMSIG_SMART_OBJECTIVE_ACHIEVED, uncompleted)
+
+/datum/objective/smart/proc/uncomplete_objective()
+	if(!completed)
+		return
+	uncompleted = TRUE
+	completed = FALSE
+	SEND_SIGNAL(src, COMSIG_SMART_OBJECTIVE_UNACHIEVED)
+
+/**
+ * ## Smart destroy AI
+ *
+ * It achieves when the AI is sent into deep space, or killed.
+ * It unachieved when the AI is back into local space, or revived.
+ */
+/datum/objective/smart/destroy_ai
+	name = "destroy AI"
+	martyr_compatible = TRUE
+	black_telecrystal_reward = 8
+
+/datum/objective/destroy/update_explanation_text()
+	..()
+	if(target?.current)
+		explanation_text = "Destroy [target.name], the experimental AI."
+	else
+		explanation_text = "Free Objective"
+
+/datum/objective/smart/destroy_ai/find_target(dupe_search_range)
+	var/list/possible_targets = active_ais(check_mind = TRUE)
+	var/mob/living/silicon/ai/target_ai = pick(possible_targets)
+	target = target_ai.mind
+	update_explanation_text()
+	return target
+
+/datum/objective/smart/destroy_ai/post_find_target()
+	if(!target || !target.current)
+		return
+	RegisterSignal(target.current, COMSIG_LIVING_DEATH, .proc/on_death)
+	RegisterSignal(target.current, COMSIG_LIVING_REVIVE, .proc/on_revive)
+	RegisterSignal(target.current, COMSIG_MOVABLE_Z_CHANGED, .proc/on_z_level_changed)
+
+/datum/objective/smart/destroy_ai/Destroy(force)
+	. = ..()
+	UnregisterSignal(target.current, list(COMSIG_LIVING_DEATH, COMSIG_LIVING_REVIVE, COMSIG_MOVABLE_Z_CHANGED))
+
+/datum/objective/smart/destroy_ai/proc/on_revive(datum/target, full_heal)
+	SIGNAL_HANDLER
+
+	uncomplete_objective()
+
+/datum/objective/smart/destroy_ai/proc/on_death(datum/target, gibbed)
+	SIGNAL_HANDLER
+
+	complete_objective()
+
+/datum/objective/smart/destroy_ai/proc/on_z_level_changed(datum/target, old_z, new_z)
+	SIGNAL_HANDLER
+
+	if(new_z > 6)
```

Comment:
🤨 

I know you didn't write this because I had to fix the same thing when rewriting comms consoles. Old coders really be wildin'.